kb/2024-05-17/modify_cloud_credential_ttl.md

```diff
@@ -0,0 +1,78 @@
+---
+title: Modifying TTL for Harvester Cloud Credential
+description: How to modify TTL for newly imported Harvester cloud credentials in Rancher 2.8.x.
+slug: modify_harvester_cloud_credential_ttl
+authors:
+  - name: Gaurav Mehta
+    title: Staff Software Engineer
+    url: https://github.com/ibrokethecloud
+    image_url: https://github.com/ibrokethecloud.png
+tags: [harvester, cloud credentials, rancher]
+hide_table_of_contents: false
+---
+
+## Kubeconfig token expiry in Rancher 2.8.x
+
+Rancher 2.8.x has introduced a new [kubeconfig-default-token-ttl-minutes](https://ranchermanager.docs.rancher.com/api/api-tokens#kubeconfig-default-token-ttl-minutes) setting. As highlighted in the document the default value for this setting is 30 days.
+
+A side effect of this issue, has been the expiry of kubeconfig tokens used by rancher in provisioning downstream clusters on Harvester.
+
+When the cloud credential token expires, the end users cannot perform any further cluster management operations on downstream clusters on Harvester. An issue tracking the same is available [here.](https://github.com/rancher/rancher/issues/44912)
+
+**NOTE:** This only impacts Harvester cloud credentials created after installing or upgrading to Rancher 2.8.x
+
+## Workaround
```

Comment:
Overall, I'm a bit concerned about this workaround. Manually editing CRDs that haven't been migrated to the RK API is risky at best, and can produce undesired side-effects. Rather than re-editing the current token, I'm thinking that the better idea would be to generate a new kubeconfig and update the cloud credential accordingly. What do you think?

Reply:
I concur with the concerns here, we should refrain from recommending users to manually update live (and privileged) credentials in production - this won't float in highly regulated sectors.

While a permanent solution is not available, I would recommend some automation that rotates the target secret every so often. That way we enforce TTL, while also removing the need of manual intervention.

We could use the Kubernetes native [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) object, to run the scripts already contained within this KB. The CronJob would be tied to a `ServiceAccount` that only has the required permissions to do the token rotation - ideally bound to the precise resource names (if possible) it needs accessing/amending.

Reply:
the bigger issue as @m-ildefons also pointed out in his work around is that if we create a new credential, the name is dynamically generated, and users will need to go and cycle through all their provisioned `cluster.provisioning` crds and patch the `.spec.cloudCredentialSecretName` to the new cloud credential.

Reply:
@ibrokethecloud I'm not sure that I understand your objection. The suggestion above is to update the cloud credential with a new kubeconfig - this would leave the cloud credential with the same name. Would you mind explaining your reasoning a bit more?

Reply:
If I may add, the workaround of manually editing some fields should only be used in case nothing else works, because there are risks as mentioned by Michael and Paulo. Otherwise, generating a new kubeconfing and updating it in the credential should be the best option at the moment.

Perhaps, a new token type in Rancher with its own TTL setting for such scenarios, e.g., integration between services, should be evaluated as a proper long term solution.

There is a security reason for having an expiring TTL set, so we can avoid proliferation of long lived unused tokens, but there is also a valid use case for such long lived tokens for integrations, as I mentioned above.